packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx

```diff
@@ -516,6 +520,46 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
     return rows;
   };
 
+  const { components, componentsProps } = rootProps;
```

Comment:
I would prefer to remove the overlays logic from the virtual scroller hook if possible, as they are not very related and the hook itself is pretty big already.
Maybe something like this would work?

```tsx
// DataGridVirtualScroller.tsx

// Handles all 3 overlays: loading, noRows, and noResults
// It combines `GridOverlays` and `GridEmptyStateOverlay` and replaces `getEmptyStateOverlay` from virtualScroller
const overlay: { type: string, element: JSX.Element } | null = useGridOverlay();

const { /* ... */ } = useGridVirtualScroller({
  ref,
  disableVirtualization,
  // this boolean should be enough to adjust styles accordingly
  isEmptyStateOverlayRendered: overlay !== null && overlay.type !== 'LoadingOverlay',
});

// ...

return (
  <GridVirtualScroller className={className} {...getRootProps(other)}>
    {overlay?.element}
    <GridVirtualScrollerContent {...getContentProps()}>


```

What do you think?